File: sdk/unreal/Source/ChillObservability/Private/ChillTypes.cpp
// Copyright Chill Contributors. Licensed under the Apache License, Version 2.0.

#include "ChillTypes.h"

namespace
{
	constexpr int32 MaximumNameLength = 128;
	constexpr int32 MaximumStringLength = 256;
	constexpr int32 MaximumArrayLength = 32;

	bool IsLowerAlpha(TCHAR Character)
	{
		return Character >= TEXT('a') && Character <= TEXT('z');
	}

	bool IsDigit(TCHAR Character)
	{
		return Character >= TEXT('0') && Character <= TEXT('9');
	}

	/** Matches ^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*$ without pulling in a regex engine. */
	bool MatchesSemantic(const FString& Value)
	{
		const int32 Length = Value.Len();
		if (Length == 0 || !IsLowerAlpha(Value[0]))
		{
			return false;
		}
		bool bPreviousWasSeparator = false;
		for (int32 Index = 1; Index < Length; ++Index)
		{
			const TCHAR Character = Value[Index];
			const bool bSeparator = Character == TEXT('.') || Character == TEXT('_') || Character == TEXT('-');
			if (bSeparator)
			{
				// A separator may never lead, trail, or repeat.
				if (bPreviousWasSeparator || Index == Length - 1)
				{
					return false;
				}
				bPreviousWasSeparator = true;
				continue;
			}
			if (!IsLowerAlpha(Character) && !IsDigit(Character))
			{
				return false;
			}
			bPreviousWasSeparator = false;
		}
		return true;
	}

	/** Matches ^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)*$. */
	bool MatchesAnnotationKey(const FString& Value)
	{
		const int32 Length = Value.Len();
		if (Length == 0)
		{
			return false;
		}
		bool bSegmentStart = true;
		for (int32 Index = 0; Index < Length; ++Index)
		{
			const TCHAR Character = Value[Index];
			if (Character == TEXT('.'))
			{
				// A dot may never lead, trail, or repeat.
				if (bSegmentStart || Index == Length - 1)
				{
					return false;
				}
				bSegmentStart = true;
				continue;
			}
			if (bSegmentStart)
			{
				if (!IsLowerAlpha(Character))
				{
					return false;
				}
				bSegmentStart = false;
				continue;
			}
			if (!IsLowerAlpha(Character) && !IsDigit(Character) && Character != TEXT('_'))
			{
				return false;
			}
		}
		return true;
	}
}

FChillDiagnostic::FChillDiagnostic(FString InCode, FString InDetail)
	: Code(MoveTemp(InCode))
	, Detail(MoveTemp(InDetail))
	, Timestamp(FDateTime::UtcNow())
{
}

namespace ChillNames
{
	bool IsSemantic(const FString& Value)
	{
		return Value.Len() <= MaximumNameLength && MatchesSemantic(Value);
	}

	bool IsAnnotationKey(const FString& Value)
	{
		return Value.Len() <= MaximumNameLength && MatchesAnnotationKey(Value);
	}

	FString SnakeCase(EChillActionActivation Value)
	{
		switch (Value)
		{
		case EChillActionActivation::Primary: return TEXT("primary");
		case EChillActionActivation::Submit: return TEXT("submit");
		case EChillActionActivation::Toggle: return TEXT("toggle");
		case EChillActionActivation::Selection: return TEXT("selection");
		case EChillActionActivation::Adjust: return TEXT("adjust");
		case EChillActionActivation::Gesture: return TEXT("gesture");
		case EChillActionActivation::System: return TEXT("system");
		}
		return TEXT("primary");
	}

	FString SnakeCase(EChillInput Value)
	{
		switch (Value)
		{
		case EChillInput::Touch: return TEXT("touch");
		case EChillInput::Pointer: return TEXT("pointer");
		case EChillInput::Keyboard: return TEXT("keyboard");
		case EChillInput::Remote: return TEXT("remote");
		case EChillInput::Accessibility: return TEXT("accessibility");
		case EChillInput::Voice: return TEXT("voice");
		case EChillInput::System: return TEXT("system");
		case EChillInput::Unknown: return TEXT("unknown");
		}
		return TEXT("unknown");
	}

	FString SnakeCase(EChillActivityKind Value)
	{
		switch (Value)
		{
		case EChillActivityKind::Ui: return TEXT("ui");
		case EChillActivityKind::Domain: return TEXT("domain");
		case EChillActivityKind::Network: return TEXT("network");
		case EChillActivityKind::Storage: return TEXT("storage");
		case EChillActivityKind::Task: return TEXT("task");
		case EChillActivityKind::Custom: return TEXT("custom");
		}
		return TEXT("domain");
	}

	FString SnakeCase(EChillEventClass Value)
	{
		switch (Value)
		{
		case EChillEventClass::Lifecycle: return TEXT("lifecycle");
		case EChillEventClass::Domain: return TEXT("domain");
		case EChillEventClass::Error: return TEXT("error");
		case EChillEventClass::Crash: return TEXT("crash");
		case EChillEventClass::Performance: return TEXT("performance");
		case EChillEventClass::Experiment: return TEXT("experiment");
		case EChillEventClass::Custom: return TEXT("custom");
		}
		return TEXT("domain");
	}

	FString SnakeCase(EChillEventSeverity Value)
	{
		switch (Value)
		{
		case EChillEventSeverity::Trace: return TEXT("trace");
		case EChillEventSeverity::Debug: return TEXT("debug");
		case EChillEventSeverity::Info: return TEXT("info");
		case EChillEventSeverity::Warn: return TEXT("warn");
		case EChillEventSeverity::Error: return TEXT("error");
		case EChillEventSeverity::Fatal: return TEXT("fatal");
		}
		return TEXT("info");
	}

	FString SnakeCase(EChillPageRelation Value)
	{
		switch (Value)
		{
		case EChillPageRelation::Root: return TEXT("root");
		case EChillPageRelation::Push: return TEXT("push");
		case EChillPageRelation::Tab: return TEXT("tab");
		case EChillPageRelation::Split: return TEXT("split");
		case EChillPageRelation::Sheet: return TEXT("sheet");
		case EChillPageRelation::Popover: return TEXT("popover");
		case EChillPageRelation::Overlay: return TEXT("overlay");
		case EChillPageRelation::Cover: return TEXT("cover");
		}
		return TEXT("root");
	}

	FString SnakeCase(EChillPageCause Value)
	{
		switch (Value)
		{
		case EChillPageCause::Initial: return TEXT("initial");
		case EChillPageCause::Navigate: return TEXT("navigate");
		case EChillPageCause::Back: return TEXT("back");
		case EChillPageCause::Selection: return TEXT("selection");
		case EChillPageCause::Present: return TEXT("present");
		case EChillPageCause::Dismiss: return TEXT("dismiss");
		case EChillPageCause::Replace: return TEXT("replace");
		case EChillPageCause::DeepLink: return TEXT("deep_link");
		case EChillPageCause::Restore: return TEXT("restore");
		case EChillPageCause::Adaptive: return TEXT("adaptive");
		case EChillPageCause::Background: return TEXT("background");
		case EChillPageCause::Foreground: return TEXT("foreground");
		case EChillPageCause::SurfaceDestroyed: return TEXT("surface_destroyed");
		}
		return TEXT("navigate");
	}

	FString SnakeCase(EChillOutcome Value)
	{
		switch (Value)
		{
		case EChillOutcome::Ok: return TEXT("ok");
		case EChillOutcome::Error: return TEXT("error");
		case EChillOutcome::Cancelled: return TEXT("cancelled");
		case EChillOutcome::Timeout: return TEXT("timeout");
		}
		return TEXT("ok");
	}

	FString SnakeCase(EChillAnnotationClassification Value)
	{
		switch (Value)
		{
		case EChillAnnotationClassification::Public: return TEXT("public");
		case EChillAnnotationClassification::Internal: return TEXT("internal");
		case EChillAnnotationClassification::PseudonymousIdentifier: return TEXT("pseudonymous_identifier");
		}
		return TEXT("internal");
	}
}

bool FChillValue::FromString(const FString& Value, FChillValue& OutValue)
{
	if (Value.Len() > MaximumStringLength)
	{
		return false;
	}
	OutValue = FChillValue();
	OutValue.Kind = EChillValueKind::String;
	OutValue.StringValue = Value;
	return true;
}

bool FChillValue::FromBoolean(bool Value, FChillValue& OutValue)
{
	OutValue = FChillValue();
	OutValue.Kind = EChillValueKind::Boolean;
	OutValue.BooleanValue = Value;
	return true;
}

bool FChillValue::FromNumber(double Value, FChillValue& OutValue)
{
	if (FMath::IsNaN(Value) || !FMath::IsFinite(Value))
	{
		return false;
	}
	OutValue = FChillValue();
	OutValue.Kind = EChillValueKind::Number;
	OutValue.NumberValue = Value;
	return true;
}

bool FChillValue::FromStrings(const TArray<FString>& Values, FChillValue& OutValue)
{
	if (Values.Num() > MaximumArrayLength)
	{
		return false;
	}
	for (const FString& Value : Values)
	{
		if (Value.Len() > MaximumStringLength)
		{
			return false;
		}
	}
	OutValue = FChillValue();
	OutValue.Kind = EChillValueKind::Strings;
	OutValue.StringValues = Values;
	return true;
}

bool FChillValue::FromBooleans(const TArray<bool>& Values, FChillValue& OutValue)
{
	if (Values.Num() > MaximumArrayLength)
	{
		return false;
	}
	OutValue = FChillValue();
	OutValue.Kind = EChillValueKind::Booleans;
	OutValue.BooleanValues = Values;
	return true;
}

bool FChillValue::FromNumbers(const TArray<double>& Values, FChillValue& OutValue)
{
	if (Values.Num() > MaximumArrayLength)
	{
		return false;
	}
	for (const double Value : Values)
	{
		if (FMath::IsNaN(Value) || !FMath::IsFinite(Value))
		{
			return false;
		}
	}
	OutValue = FChillValue();
	OutValue.Kind = EChillValueKind::Numbers;
	OutValue.NumberValues = Values;
	return true;
}

FChillAnnotations FChillAnnotations::With(const FString& Key, const FString& Value) const
{
	FChillValue Converted;
	return FChillValue::FromString(Value, Converted) ? Add(Key, Converted) : *this;
}

FChillAnnotations FChillAnnotations::With(const FString& Key, bool Value) const
{
	FChillValue Converted;
	return FChillValue::FromBoolean(Value, Converted) ? Add(Key, Converted) : *this;
}

FChillAnnotations FChillAnnotations::With(const FString& Key, double Value) const
{
	FChillValue Converted;
	return FChillValue::FromNumber(Value, Converted) ? Add(Key, Converted) : *this;
}

FChillAnnotations FChillAnnotations::With(const FString& Key, const TArray<FString>& Values) const
{
	FChillValue Converted;
	return FChillValue::FromStrings(Values, Converted) ? Add(Key, Converted) : *this;
}

FChillAnnotations FChillAnnotations::With(const FString& Key, const TArray<bool>& Values) const
{
	FChillValue Converted;
	return FChillValue::FromBooleans(Values, Converted) ? Add(Key, Converted) : *this;
}

FChillAnnotations FChillAnnotations::With(const FString& Key, const TArray<double>& Values) const
{
	FChillValue Converted;
	return FChillValue::FromNumbers(Values, Converted) ? Add(Key, Converted) : *this;
}

FChillAnnotations FChillAnnotations::MergeDescendant(const FChillAnnotations& Descendant) const
{
	FChillAnnotations Merged = *this;
	for (const FChillAnnotationEntry& Candidate : Descendant.Entries)
	{
		if (!Merged.Contains(Candidate.Key))
		{
			Merged.Entries.Add(Candidate);
		}
	}
	return Merged;
}

FChillAnnotations FChillAnnotations::Add(const FString& Key, const FChillValue& Value) const
{
	if (!ChillNames::IsAnnotationKey(Key) || Contains(Key))
	{
		return *this;
	}
	FChillAnnotations Copied = *this;
	Copied.Entries.Emplace(Key, Value);
	return Copied;
}

bool FChillAnnotations::Contains(const FString& Key) const
{
	for (const FChillAnnotationEntry& Entry : Entries)
	{
		if (Entry.Key.Equals(Key, ESearchCase::CaseSensitive))
		{
			return true;
		}
	}
	return false;
}

File: sdk/unity/Runtime/ChillTypes.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text.RegularExpressions;

namespace Chill.Unity
{
    public enum ChillConsent
    {
        Denied,
        Granted
    }

    public enum ChillAnnotationClassification
    {
        Public,
        Internal,
        PseudonymousIdentifier
    }

    public enum ChillActionActivation
    {
        Primary,
        Submit,
        Toggle,
        Selection,
        Adjust,
        Gesture,
        System
    }

    public enum ChillInput
    {
        Touch,
        Pointer,
        Keyboard,
        Remote,
        Accessibility,
        Voice,
        System,
        Unknown
    }

    public enum ChillActivityKind
    {
        Ui,
        Domain,
        Network,
        Storage,
        Task,
        Custom
    }

    public enum ChillEventClass
    {
        Lifecycle,
        Domain,
        Error,
        Crash,
        Performance,
        Experiment,
        Custom
    }

    public enum ChillEventSeverity
    {
        Trace,
        Debug,
        Info,
        Warn,
        Error,
        Fatal
    }

    public enum ChillPageRelation
    {
        Root,
        Push,
        Tab,
        Split,
        Sheet,
        Popover,
        Overlay,
        Cover
    }

    public enum ChillPageCause
    {
        Initial,
        Navigate,
        Back,
        Selection,
        Present,
        Dismiss,
        Replace,
        DeepLink,
        Restore,
        Adaptive,
        Background,
        Foreground,
        SurfaceDestroyed
    }

    public enum ChillOutcome
    {
        Ok,
        Error,
        Cancelled,
        Timeout
    }

    public sealed class ChillDiagnostic
    {
        internal ChillDiagnostic(string code, string detail)
        {
            Code = code;
            Detail = detail;
            Timestamp = DateTimeOffset.UtcNow;
        }

        public string Code { get; private set; }
        public string Detail { get; private set; }
        public DateTimeOffset Timestamp { get; private set; }
    }

    internal static class ChillNames
    {
        private static readonly Regex SemanticName = new Regex(
            "^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*$",
            RegexOptions.CultureInvariant);

        private static readonly Regex AnnotationKey = new Regex(
            "^[a-z][a-z0-9_]*(?:\\.[a-z][a-z0-9_]*)*$",
            RegexOptions.CultureInvariant);

        internal static bool IsSemantic(string value)
        {
            return value != null && value.Length <= 128 && SemanticName.IsMatch(value);
        }

        internal static bool IsAnnotationKey(string value)
        {
            return value != null && value.Length <= 128 && AnnotationKey.IsMatch(value);
        }

        internal static string SnakeCase(Enum value)
        {
            string source = value.ToString();
            var result = new System.Text.StringBuilder(source.Length + 4);
            for (int index = 0; index < source.Length; index += 1)
            {
                char current = source[index];
                if (char.IsUpper(current) && index > 0)
                {
                    result.Append('_');
                }
                result.Append(char.ToLowerInvariant(current));
            }
            return result.ToString();
        }
    }

    internal enum ChillValueKind
    {
        String,
        Boolean,
        Number,
        Strings,
        Booleans,
        Numbers
    }

    internal sealed class ChillValue
    {
        private ChillValue(ChillValueKind kind, object value)
        {
            Kind = kind;
            Value = value;
        }

        internal ChillValueKind Kind { get; private set; }
        internal object Value { get; private set; }

        internal static ChillValue From(string value)
        {
            if (value == null || value.Length > 256)
            {
                throw new ArgumentException("String annotations must contain at most 256 characters.");
            }
            return new ChillValue(ChillValueKind.String, value);
        }

        internal static ChillValue From(bool value)
        {
            return new ChillValue(ChillValueKind.Boolean, value);
        }

        internal static ChillValue From(double value)
        {
            if (double.IsNaN(value) || double.IsInfinity(value))
            {
                throw new ArgumentException("Number annotations must be finite.");
            }
            return new ChillValue(ChillValueKind.Number, value);
        }

        internal static ChillValue From(IReadOnlyList<string> values)
        {
            if (values == null || values.Count > 32)
            {
                throw new ArgumentException("Annotation arrays must contain at most 32 values.");
            }
            var copied = new string[values.Count];
            for (int index = 0; index < values.Count; index += 1)
            {
                if (values[index] == null || values[index].Length > 256)
                {
                    throw new ArgumentException("String annotations must contain at most 256 characters.");
                }
                copied[index] = values[index];
            }
            return new ChillValue(ChillValueKind.Strings, copied);
        }

        internal static ChillValue From(IReadOnlyList<bool> values)
        {
            if (values == null || values.Count > 32)
            {
                throw new ArgumentException("Annotation arrays must contain at most 32 values.");
            }
            var copied = new bool[values.Count];
            for (int index = 0; index < values.Count; index += 1)
            {
                copied[index] = values[index];
            }
            return new ChillValue(ChillValueKind.Booleans, copied);
        }

        internal static ChillValue From(IReadOnlyList<double> values)
        {
            if (values == null || values.Count > 32)
            {
                throw new ArgumentException("Annotation arrays must contain at most 32 values.");
            }
            var copied = new double[values.Count];
            for (int index = 0; index < values.Count; index += 1)
            {
                if (double.IsNaN(values[index]) || double.IsInfinity(values[index]))
                {
                    throw new ArgumentException("Number annotations must be finite.");
                }
                copied[index] = values[index];
            }
            return new ChillValue(ChillValueKind.Numbers, copied);
        }
    }

    internal sealed class ChillAnnotationEntry
    {
        internal ChillAnnotationEntry(string key, ChillValue value)
        {
            Key = key;
            Value = value;
        }

        internal string Key { get; private set; }
        internal ChillValue Value { get; private set; }
    }

    public sealed class ChillAnnotations
    {
        private readonly List<ChillAnnotationEntry> entries;

        public ChillAnnotations()
        {
            entries = new List<ChillAnnotationEntry>();
        }

        private ChillAnnotations(List<ChillAnnotationEntry> entries)
        {
            this.entries = entries;
        }

        internal IReadOnlyList<ChillAnnotationEntry> Entries
        {
            get { return new ReadOnlyCollection<ChillAnnotationEntry>(entries); }
        }

        public ChillAnnotations With(string key, string value)
        {
            return Add(key, ChillValue.From(value));
        }

        public ChillAnnotations With(string key, bool value)
        {
            return Add(key, ChillValue.From(value));
        }

        public ChillAnnotations With(string key, double value)
        {
            return Add(key, ChillValue.From(value));
        }

        public ChillAnnotations With(string key, IReadOnlyList<string> values)
        {
            return Add(key, ChillValue.From(values));
        }

        public ChillAnnotations With(string key, IReadOnlyList<bool> values)
        {
            return Add(key, ChillValue.From(values));
        }

        public ChillAnnotations With(string key, IReadOnlyList<double> values)
        {
            return Add(key, ChillValue.From(values));
        }

        public ChillAnnotations MergeDescendant(ChillAnnotations descendant)
        {
            if (descendant == null)
            {
                return this;
            }
            var merged = new List<ChillAnnotationEntry>(entries);
            for (int index = 0; index < descendant.entries.Count; index += 1)
            {
                ChillAnnotationEntry candidate = descendant.entries[index];
                if (!Contains(merged, candidate.Key))
                {
                    merged.Add(candidate);
                }
            }
            return new ChillAnnotations(merged);
        }

        private ChillAnnotations Add(string key, ChillValue value)
        {
            if (!ChillNames.IsAnnotationKey(key))
            {
                throw new ArgumentException("Annotation keys must be lowercase dotted identifiers.", "key");
            }
            if (Contains(entries, key))
            {
                return this;
            }
            var copied = new List<ChillAnnotationEntry>(entries);
            copied.Add(new ChillAnnotationEntry(key, value));
            return new ChillAnnotations(copied);
        }

        private static bool Contains(List<ChillAnnotationEntry> values, string key)
        {
            for (int index = 0; index < values.Count; index += 1)
            {
                if (string.Equals(values[index].Key, key, StringComparison.Ordinal))
                {
                    return true;
                }
            }
            return false;
        }
    }
}
